components/lib/dataprotect/src/main/java/mozilla/components/lib/dataprotect/Keystore.kt

```diff
@@ -0,0 +1,98 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.dataprotect
+
+import android.annotation.TargetApi
+import android.os.Build
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyProperties
+import java.security.KeyStore
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.SecretKey
+import javax.crypto.spec.GCMParameterSpec
+
+private val KEYSTORE_PROVIDER = "AndroidKeyStore"
+private val ENCRYPTED_VERSION = 0x02
+
+private val CIPHER_ALG = KeyProperties.KEY_ALGORITHM_AES
+private val CIPHER_MOD = KeyProperties.BLOCK_MODE_GCM
+private val CIPHER_PAD = KeyProperties.ENCRYPTION_PADDING_NONE
+private val CIPHER_TAG_LEN = 128
+private val CIPHER_SPEC = "${CIPHER_ALG}/${CIPHER_MOD}/${CIPHER_PAD}"
+
+@TargetApi(Build.VERSION_CODES.M)
+class Keystore {
```

Comment:
I wonder if the API would be simpler if `label` would be a constructor parameter of `KeyStore` and you'd create multiple instances if you'd need multiple labels. But that may depend on how this is going to be used in an app, I guess.

Reply:
I was basing the API on work done in NSS for [OSKeyStore](https://dxr.mozilla.org/mozilla-central/source/security/manager/ssl/nsIOSKeyStore.idl).  We don't have to be religious about it, and I suspect the work is about the same for both models.

In Lockbox, I think we'll just use a single label.  I can foresee how multiple labels could be applied.

I'm happy to change this to a single label set on construction if that's the preference.

Reply:
If there will be only a single label in lockbox then the constructor API may be more handy not requiring you to pass the label around all the time? :)

But yeah, not religious about it either. :)

Reply:
That's the only thing I noticed as well as the `generateKey` method seemed suspicious to me, since it's not returning anything. Not a big deal, but would also prefer the constructor parameter and `Keystore` being immutable.

Reply:
I do need to note the exceptions to be thrown, which are a lot \-:

Automatically creating the key risks some data corruption and crashing if the keystore  was reset for some reason but the data stored did not; hence the separation.

Definitely open to suggestions, though.